Adding the actual mining process to the recipes (basic-pow, hybrid-consensus)

Mhnd3 opened the pull request.

Mhnd3 committed:
Update service.rs

Add the actual mining process.

Mhnd3 committed:
Update service.rs

Add the actual mining process.

Mhnd3 committed:
minor update service.rs

Mhnd3 committed:
Update pow consensus lib.rs

- public "hash_meets_difficulty" function.
- reduce difficulty due to the use of u8 type as nonce for pow mining.

jimmychu0807 commented on consensus/sha3pow/src/lib.rs:
```diff
@@ -60,7 +60,7 @@ impl<B: BlockT<Hash = H256>> PowAlgorithm<B> for MinimalSha3Algorithm {
 	fn difficulty(&self, _parent: B::Hash) -> Result<Self::Difficulty, Error<B>> {
 		// Fixed difficulty hardcoded here
-		Ok(U256::from(1_000_000))
+		Ok(U256::from(1_0))
```
```suggestion
		Ok(U256::from(10))
```

jimmychu0807 reviewed (dismissed):
@Mhnd3 Thanks for the fix on making basic-pow and hybrid-consensus node start producing blocks again. 

@JoshOrndorff @Mhnd3 there are some questions I am not sure about in the PoW code, raised in the code inline. Could you shed me some light here? Thanks.

jimmychu0807 committed:
Update consensus/sha3pow/src/lib.rs

jimmychu0807 commented on nodes/basic-pow/src/service.rs:
```diff
@@ -193,6 +196,33 @@ pub fn new_full(config: Configuration) -> Result<TaskManager, ServiceError> {
+				let worker = _worker.clone();
+				let metadata = worker.lock().metadata();
+				if let Some(metadata) = metadata {
+					let nonce = H256::from_slice(Sha3_256::digest(&[numb]).as_slice());
```
Actually I am not familiar with writing consensus, so I am also learning here. A few questions here:

Why is `nonce` a hash, instead of simply an incrementing number? I see that numb pretty much act as an incrementing nonce. But then this number should probably be of type `U256` instead of `u8`.

Mhnd3 replied:
Thanks @jimmychu0807 for the review!

The `nonce` as a Hash is mimicking the Kulupu implementation that uses the RandomX algorithm, as one of it's security goals that the Miners should not be able to identify "weaker" authorised parameters (such as specific nonce values) that make the PoW computation significantly more efficient, thus giving an advantage to a miner who would know such parameters, and in Kulupu implementation the `nonce` is actully the Hash of a random value 
`let nonce = H256::random_using(&mut rng);`
In this implementation it doesn't really need to be a Hash at all, I think an `U256` would be a better fit, as we really do increment the `nonce` by `1` every unsuccessful attempt.
the `u8` was used cuz i didn't change anything in the rest of the code, where the `nonce` is a Hash, and the `Sha3_256::digest()` takes a `&[u8]` as input, so i found it a faster way.
Bottom line using the `U256` for the nonce would eliminate the need for the `numb` variable. It was just quick and dirty implementation to make it work as discussed in https://github.com/substrate-developer-hub/recipes/issues/432 
I updated the code to make it uses the type `U256` for the nonce.

jimmychu0807 commented on nodes/basic-pow/src/service.rs:
```diff
@@ -193,6 +196,33 @@ pub fn new_full(config: Configuration) -> Result<TaskManager, ServiceError> {
+					if numb == 255u8 {
+						numb = 0;
+					}
+					thread::sleep(Duration::new(0, 500_000_000));
```
Isn't the logic here should be something like this?

```suggestion
					if hash_meets_difficulty(&seal.work, seal.difficulty) {
						let mut worker = worker.lock();
						worker.submit(seal.encode());
					} else {
					  numb = numb.saturating_add(1u8);
					  if numb == 255u8 {
						  numb = 0;
					  }
					}
```

The last line of `thread::sleep` is in the `else` branch of `if let Some(metadata) = metadata {..} else { thread::sleep }`, based on code structure in https://github.com/kulupu/kulupu/blob/master/src/service.rs#L330-L377

Mhnd3 replied:
putting the `thread::sleep` in the `else` branch is for producing a delay in case no Block is yet ready to start mining, but i added it to create a delay in the mining process itself, since the `difficulty` was reduced due to the usage of the `u8` numb variable to generate a nonce.
i updated this, the nonce uses the type `U256` now.

Mhnd3 committed:
Using the `U256` type for nonce

- since there is no need for a random nonce, the nonce type `H256` was replaced with `U256`.
- minor changes.

Mhnd3 committed:
Using the `U256` type for nonce

change the nonce type from `H256` to `U256`, as a Hash is not needed as nonce for this implementation.

Mhnd3 committed:
Using the `U256` type for nonce

jimmychu0807 commented on nodes/basic-pow/src/service.rs:
```diff
@@ -193,6 +195,33 @@ pub fn new_full(config: Configuration) -> Result<TaskManager, ServiceError> {
+		    } else {
+			thread::sleep(Duration::new(1, 0));
+		    }
+		});
```
```suggestion
		let mut nonce: U256 = U256::from(0);
		thread::spawn(move || loop {
			let worker = _worker.clone();
			let metadata = worker.lock().metadata();
			if let Some(metadata) = metadata {
				let compute = Compute {
					difficulty: metadata.difficulty,
					pre_hash: metadata.pre_hash,
					nonce,
				};
				let seal = compute.compute();
				if hash_meets_difficulty(&seal.work, seal.difficulty) {
					nonce = U256::from(0);
					let mut worker = worker.lock();
					worker.submit(seal.encode());
				} else {
					nonce = nonce.saturating_add(U256::from(1));
					if nonce == U256::MAX {
						nonce = U256::from(0);
					}
				}
			} else {
				thread::sleep(Duration::new(1, 0));
			}
		});
```

jimmychu0807 replied:
Unifying indenting spaces to indenting tabs

jimmychu0807 commented on nodes/hybrid-consensus/src/service.rs:
```diff
@@ -206,6 +208,33 @@ pub fn new_full(config: Configuration) -> Result<TaskManager, ServiceError> {
+        	    } else {
+        	        thread::sleep(Duration::new(1, 0));
+        	    }
+		});
```
```suggestion
		// Start Mining
		let mut nonce: U256 = U256::from(0);
		thread::spawn(move || loop {
			let worker = _worker.clone();
			let metadata = worker.lock().metadata();
			if let Some(metadata) = metadata {
				let compute = Compute {
					difficulty: metadata.difficulty,
					pre_hash: metadata.pre_hash,
					nonce,
				};
				let seal = compute.compute();
				if hash_meets_difficulty(&seal.work, seal.difficulty) {
					nonce = U256::from(0);
					let mut worker = worker.lock();
					worker.submit(seal.encode());
				} else {
					nonce = nonce.saturating_add(U256::from(1));
					if nonce == U256::MAX {
						nonce = U256::from(0);
					}
				}
			} else {
				thread::sleep(Duration::new(1, 0));
			}
		});
```

jimmychu0807 replied:
Unifying all indenting spaces to indenting tabs

jimmychu0807 committed:
Update nodes/hybrid-consensus/src/service.rs

jimmychu0807 committed:
Update nodes/basic-pow/src/service.rs

jimmychu0807 commented on nodes/hybrid-consensus/src/service.rs:
```diff
@@ -6,12 +6,14 @@ use sc_executor::native_executor_instance;
 use std::sync::Arc;
 use std::time::Duration;
+use std::thread;
+use sp_core::{H256, Encode};
```
```suggestion
use sp_core::{U256, Encode};
```

jimmychu0807 committed:
Update nodes/hybrid-consensus/src/service.rs
